src/daft-json/src/decoding.rs

```diff
@@ -202,189 +195,81 @@ fn deserialize_into<'a, A: Borrow<Value<'a>>>(target: &mut Box<dyn MutableArray>
     }
 }
 
-fn deserialize_primitive_into<'a, A: Borrow<Value<'a>>, T: NativeType>(
+fn deserialize_primitive_into<'a, A: Borrow<BorrowedValue<'a>>, T: NativeType + NumCast>(
     target: &mut Box<dyn MutableArray>,
     rows: &[A],
-    deserialize_into: fn(&mut MutablePrimitiveArray<T>, &[A]) -> (),
 ) {
-    generic_deserialize_into(target, rows, deserialize_into)
+    let target = target
+        .as_mut_any()
+        .downcast_mut::<MutablePrimitiveArray<T>>()
+        .unwrap();
+
+    let iter = rows.iter().map(|row| match row.borrow() {
+        BorrowedValue::Static(StaticNode::I64(v)) => T::from(*v),
+        BorrowedValue::Static(StaticNode::U64(v)) => T::from(*v),
+        BorrowedValue::Static(StaticNode::F64(v)) => T::from(*v),
+        BorrowedValue::Static(StaticNode::Bool(v)) => T::from(*v as u8),
+        _ => None,
+    });
+    target.extend_trusted_len(iter);
 }
 
-fn generic_deserialize_into<'a, A: Borrow<Value<'a>>, M: 'static>(
+fn generic_deserialize_into<'a, A: Borrow<BorrowedValue<'a>>, M: 'static>(
     target: &mut Box<dyn MutableArray>,
     rows: &[A],
     deserialize_into: fn(&mut M, &[A]) -> (),
 ) {
     deserialize_into(target.as_mut_any().downcast_mut::<M>().unwrap(), rows);
 }
 
-fn deserialize_utf8_into<'a, O: Offset, A: Borrow<Value<'a>>>(
+fn deserialize_utf8_into<'a, O: Offset, A: Borrow<BorrowedValue<'a>>>(
     target: &mut MutableUtf8Array<O>,
     rows: &[A],
 ) {
-    let mut scratch = vec![];
+    let mut scratch = String::new();
+
     for row in rows {
         match row.borrow() {
-            Value::String(v) => target.push(Some(v.as_ref())),
-            Value::Number(number) => match number {
-                Number::Integer(number, exponent) | Number::Float(number, exponent) => {
-                    scratch.clear();
-                    scratch.extend_from_slice(number);
-                    if !exponent.is_empty() {
-                        scratch.push(b'e');
-                        scratch.extend_from_slice(exponent);
-                    }
-                    target.push(simdutf8::basic::from_utf8(scratch.as_slice()).ok());
-                }
-            },
-            Value::Bool(v) => target.push(Some(if *v { "true" } else { "false" })),
+            BorrowedValue::String(v) => target.push(Some(v.as_ref())),
+            BorrowedValue::Static(StaticNode::Bool(v)) => {
+                target.push(Some(if *v { "true" } else { "false" }))
+            }
+            BorrowedValue::Static(node) if !matches!(node, StaticNode::Null) => {
+                write!(scratch, "{node}").unwrap();
+                target.push(Some(scratch.as_str()));
+                scratch.clear();
+            }
```

Comment:
For integers and floats that contain exponents, I'm assuming that the string representation of `StaticNode::I64` and `StaticNode::F64` values won't preserve the exponent format, right? E.g. `1.5e9`, when decoded into a string field, will be decoded as `"1500000000"`. Is that correct?

Reply:
Yeah, without manually parsing from simd_json's `Tape`, there isn't a way to preserve the exponent format. r

Reply:
I think that change in behavior is fine, we can always special-case that if it ends up being an issue.